### PR TITLE
Don't warn repeatedly for same un-optimize-able op.

### PR DESCRIPTION
**Pull request by rpgoldman**
Fix for issue #23.

Keep a set of names of ops for which "can't optimize" warnings have already been issued. Check that set (and update it, if necessary) before issuing a warning to avoid warning floods.

**Thank your for opening a PR!**

Here are a few important guidelines and requirements to check before your PR can be merged:
+ [x] There is an informative high-level description of the changes.
+ [x] The description and/or commit message(s) references the relevant GitHub issue(s).
+ [ ] The commit messages follow [these guidelines](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html).
+ [ ] The commits correspond to [relevant logical changes](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes).
+ [ ] There are tests covering the changes introduced in the PR.
+ [ ] Remove whitespace reformatting or factor into separate commit.
+ [ ] (Possibly) pull out repeated code into function.

### CODE REVIEW

**Review by twiecki — APPROVED** [review]

**Review by brandonwillard — CHANGES_REQUESTED** [review]
If we want to limit the messages output by the logger, then we should [add a filter](https://docs.python.org/3/library/logging.html#logging.Logger.addFilter).

**Comment by brandonwillard** [comment]
It looks like we already have an implementation of just such a log filter in [`theano.compile.debugmode`](https://github.com/pymc-devs/Theano-PyMC/blob/9f7cf7d09fb2139ba15a06e385dae50527614f83/theano/compile/debugmode.py#L40), so we could generalize that a little, move it to `theano.utils`, and reuse it here.

**Comment by michaelosthege** [comment]
@rpgoldman what's the state of this PR? We want to get a release going for PyMC3 and this one is relevant

**Comment by twiecki** [comment]
I think this PR needs to change to use a log-filter as requested by @brandonwillard.

**Comment by rpgoldman** [comment]
I have a draft of a new version using log filters. Haven’t had time to add tests, but hope to push it soon.

**Comment by michaelosthege** [comment]
@rpgoldman I rebased your branch. The code style should now be fixed too.
Hope this works.

**Review by twiecki — APPROVED** [review]

File: theano/compile/debugmode.py
Comment (brandonwillard) [inline theano/compile/debugmode.py:537]: Why was this added?

**Review by brandonwillard — CHANGES_REQUESTED** [review]

File: theano/compile/debugmode.py
Comment (rpgoldman) [reply theano/compile/debugmode.py:537]: I put it I to make mypy happy, to support my claim that `id_str` was a string, IIRC.

Comment (brandonwillard) [reply theano/compile/debugmode.py:537]: Don't worry about `mypy` results; this project doesn't officially support type checking, yet.

Regardless, we need to remove this before we can merge.

**Comment by michaelosthege** [comment]
This PR is looking really good. Personally I don't think that the assert is a problem or that it should block this from moving forward.
I would like to help, but I tried to rebase this branch already yesterday and failed.

We need to change something to prevent situations like this (PR blocked because of git history). Our git history commits are not flying to the Moon.

File: theano/compile/debugmode.py
Comment (rpgoldman) [reply theano/compile/debugmode.py:537]: Hi @brandonwillard -- I'm ok with doing this, but I wonder why you are so firm about it needing to be removed.  Is there some design or style criterion this violates?

Comment (rpgoldman) [reply theano/compile/debugmode.py:537]: By the way, in connection with this question, anyone know what the heck the `get_printed` argument is for?  Happy to add doc for this to the PR if someone provides an answer.

**Comment by rpgoldman** [comment]
Fixed @brandonwillard 's issue, rebased, and force-pushed. Should be good to go.

File: theano/compile/debugmode.py
Comment (brandonwillard) [reply theano/compile/debugmode.py]: I'm not familiar enough with this function to know whether or not that `assert` is valid.  For instance, it might be acceptable for `id_str` to be `None`, or perhaps we could be more explicit and avoid an `assert` altogether with an `else` statement that sets `id_str = ""`.  Plus, it's almost always better to raise a more informative `Exception` when/if you expect an expression to be violated.  (Personally, I think the former would be ideal.)

Since this change isn't blocking the PR, it's hard to justify walking through `debugmode.py`'s logic in order to answer these questions.  Also, even if the tests currently pass with this `assert`, I don't believe we're using `debugmode` much at all, so it might not actually be getting the coverage it needs.

Regarding `mypy`, can't you set `id_str: Optional[Text]`?  That would allow `id_str` to be `None`&mdash;according to `mypy`, at least.

**Comment by brandonwillard** [comment]
I just squashed and fixed the style issues.

**Pull request merged**
